src/composables/useGetParticipants.js

```diff
@@ -0,0 +1,135 @@
+	const isInCall = useIsInCall()
+	const oneToOneConversation = computed(() => conversation.value.type === CONVERSATION.TYPE.ONE_TO_ONE)
+
+	// External use
+	const participantsInitialised = ref(false)
+	const fetchingParticipants = ref(false)
+	const pendingChanges = ref(false)
+	let debounceFastUpdateParticipants = ref(null)
+	let debounceSlowUpdateParticipants = ref(null)
+
+	/**
+	 * Initialise the get participants functionality
+	 *
+	 */
+	function initialiseGetParticipants() {
+		EventBus.$on('route-change', onRouteChange)
+		EventBus.$on('joined-conversation', onJoinedConversation)
+
+		// FIXME this works only temporary until signaling is fixed to be only on the calls
+		// Then we have to search for another solution. Maybe the room list which we update
+		// periodically gets a hash of all online sessions?
+		EventBus.$on('signaling-participant-list-changed', debounceUpdateParticipants)
+		subscribe('guest-promoted', onJoinedConversation)
+	}
+
+	/**
+	 * Stop the get participants functionality
+	 *
+	 */
+	function stopGetParticipants() {
+		EventBus.$off('route-change', onRouteChange)
+		EventBus.$off('joined-conversation', onJoinedConversation)
+		EventBus.$off('signaling-participant-list-changed', debounceUpdateParticipants)
+		unsubscribe('guest-promoted', onJoinedConversation)
+	}
+
+	const onRouteChange = () => {
+		participantsInitialised.value = store.getters.participantsList(token.value).length > 1
+	}
+
+	const onJoinedConversation = () => {
+		if (oneToOneConversation.value) {
+			cancelableGetParticipants()
+		}
+		nextTick(debounceUpdateParticipants())
+	}
+
+	const debounceUpdateParticipants = () => {
+		if (!isActive && !isInCall.value) {
+			// Update is ignored but there is a flag to force the participants update
+			pendingChanges.value = true
+			return
+		}
+
+		if (store.getters.windowIsVisible() && (isInCall.value || !conversation.value?.hasCall)) {
+			debounceFastUpdateParticipants()
+		} else {
+			debounceSlowUpdateParticipants()
+		}
+		pendingChanges.value = false
+	}
+
+	const cancelableGetParticipants = async () => {
+		const isInLobby = store.getters.isInLobby
+		const isModeratorOrUser = store.getters.isModeratorOrUser
+		if (fetchingParticipants.value || token.value === '' || isInLobby || !isModeratorOrUser) {
+			return
+		}
+
+		fetchingParticipants.value = true
+		debounceFastUpdateParticipants.clear()
+		debounceSlowUpdateParticipants.clear()
+
+		const response = await store.dispatch('fetchParticipants', { token: token.value })
+		if (response) {
+			participantsInitialised.value = true
+		}
+		fetchingParticipants.value = false
+	}
+
+	onMounted(() => {
+		debounceFastUpdateParticipants = debounce(cancelableGetParticipants, 3000)
+		debounceSlowUpdateParticipants = debounce(cancelableGetParticipants, 15000)
+	})
+
+	return {
+		pendingChanges,
+		participantsInitialised,
+		initialiseGetParticipants,
+		stopGetParticipants,
+		debounceUpdateParticipants,
+		cancelableGetParticipants,
+	}
```

Comment:
Here first `return` return `debounceFastUpdateParticipants` variable with an empty `ref`.
And then in `onMounted` variable is changed to a function.

But `return` happened already, and the result of the composable is a lost empty `ref(null)`.

It would work, if the value is changed here `debounceFastUpdateParticipants.value = debounce`
But debounce function can be defined initially, there is no need to wait for mounting to DOM.

So `debounceFastUpdateParticipants` should be a const function.

Reply:
Should be fine if we'll omit `onMounted` here, and assign it during the initialization?
`debounce` here returns an object with  `.clear` property, so it's not a simple function

Reply:
> Should be fine if we'll omit `onMounted` here, and assign it during the initialization?

Yes, and it can be a `const`. We don't need to re-assign it, it is always the same function that exists initially.

> an object with `.clear` property, so it's not a simple function

Still a function =D

Functions have methods and properties (like `toString()` or `bind()`). And JS allows this:

```js
function foo() {}

foo.bar = 'bar'
foo.clear = () => {}
```

Reply:
I kept it defined like this (not a function directly)
```js
const debounceFastUpdateParticipants = debounce(
		cancelableGetParticipants, 3000)
```

Because `.clear()` is a method related to debounce and cannot be called from a wrapping function.
```js
function debounceFastUpdateParticipants() {
   debounce(cancelableGetParticipants, 3000)
}
```

But if we want to keep it as function, this means that .clear should be explicitly defined to `debounceFastUpdateParticipants` like this ? 
```js
debounceFastUpdateParticipants.clear = () => {
  debounce(cancelableGetParticipants, 3000).clear()
}
```

Reply:
This is a correct defining of a debounced function. `debounce` returns a function, so `debounceFastUpdateParticipants` is a function here.

```js
const debounceFastUpdateParticipants = debounce(
		cancelableGetParticipants, 3000)
```